stablehlo/dialect/AssemblyFormat.h

```diff
@@ -213,6 +221,90 @@ void printCustomCallTarget(AsmPrinter& p, Operation*, StringAttr target);
 
 ParseResult parseCustomCallTarget(AsmParser& parser, StringAttr& target);
 
+// DotDimensionNumbers - Abbreviated printing using a ConvDimensionNumbers-
+// inspired notation. batching_dims are skipped if empty.
+//
+// Generic:
+//    dot_dimension_numbers = #stablehlo.dot<
+//      lhs_batching_dimensions = [],
+//      lhs_contracting_dimensions = [1],
+//      rhs_batching_dimensions = [],
+//      rhs_contracting_dimensions = [0]
+//    >
+//    dot_dimension_numbers = #stablehlo.dot<
+//      lhs_batching_dimensions = [0],
+//      lhs_contracting_dimensions = [2],
+//      rhs_batching_dimensions = [0],
+//      rhs_contracting_dimensions = [1]
+//    >
+//
+// Custom:
+//    contracting_dims = [1] x [0]
+//    batching_dims = [0] x [0], contracting_dims = [2] x [1]
+template <typename AttrTy>
+void printDotDimensionNumbers(AsmPrinter& p, Operation* op, AttrTy target) {
+  // Print two ArrayRef<int64_t> as `[...] x [...]`
+  auto printLhsRhsDims = [&](ArrayRef<int64_t> lhsDims,
+                             ArrayRef<int64_t> rhsDims) {
+    DenseI64ArrayAttr::get(op->getContext(), lhsDims).print(p);
```

Comment:
I like this inline create&print. I'll look into refactoring the other DenseArray printing we have to resemble this.